In [1]:
#!/usr/bin/env python
# -*- coding: utf-8 -*-


import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.tsa.stattools import acf
from statsmodels.formula.api import ols
import matplotlib.pyplot as plt
from statsmodels.tsa.ar_model import AutoReg
from statsmodels.stats.outliers_influence import summary_table
import wooldridge

# (c1)
print("#(c1)")
print("#(i)Compute first order autocorrelation")

# Load hseinv dataset
hseinv = wooldridge.data('hseinv')
hseinv

#(c1)
#(i)Compute first order autocorrelation


year       inv       pop   price       linv       lpop    lprice   t  \
0   1947   54864.0  144126.0  0.8190  10.912613  11.878443 -0.199671   1   
1   1948   64717.0  146631.0  0.8649  11.077779  11.895675 -0.145141   2   
2   1949   63150.0  149188.0  0.8456  11.053268  11.912963 -0.167709   3   
3   1950   86014.0  151684.0  0.8765  11.362266  11.929555 -0.131819   4   
4   1951   70610.0  154287.0  0.8819  11.164927  11.946569 -0.125677   5   
5   1952   68574.0  156954.0  0.8842  11.135669  11.963708 -0.123072   6   
6   1953   70818.0  159565.0  0.8868  11.167869  11.980206 -0.120136   7   
7   1954   78460.0  162391.0  0.8597  11.270344  11.997763 -0.151172   8   
8   1955   91204.0  165275.0  0.8708  11.420854  12.015366 -0.138343   9   
9   1956   80383.0  168221.0  0.8829  11.294558  12.033034 -0.124543  10   
10  1957   74040.0  171274.0  0.8722  11.212360  12.051020 -0.136737  11   
11  1958   74822.0  174141.0  0.8521  11.222867  12.067620 -0.160051  12   
12  1959   88936.0  177830.0  0.8647  11.395672  12.088583 -0.145373  13   
13  1960   83127.0  180671.0  0.8620  11.328125  12.104433 -0.148500  14   
14  1961   83207.0  183691.0  0.8553  11.329086  12.121010 -0.156303  15   
15  1962   89121.0  186538.0  0.8593  11.397750  12.136391 -0.151637  16   
16  1963   96778.0  189242.0  0.8656  11.480175  12.150782 -0.144332  17   
17  1964   94306.0  191889.0  0.8795  11.454300  12.164672 -0.128402  18   
18  1965  100103.0  194303.0  0.8774  11.513955  12.177174 -0.130792  19   
19  1966   92145.0  196560.0  0.8783  11.431119  12.188723 -0.129767  20   
20  1967   91336.0  198712.0  0.8919  11.422300  12.199612 -0.114401  21   
21  1968   99617.0  200706.0  0.9036  11.509089  12.209597 -0.101369  22   
22  1969  102183.0  202677.0  0.9152  11.534520  12.219369 -0.088613  23   
23  1970   96855.0  205052.0  0.8823  11.480970  12.231019 -0.125223  24   
24  1971  124309.0  207191.0  0.8798  11.730526  12.241396 -0.128061  25   
25  1972  147269.0  209353.0  0.8874  11.900016  12.251777 -0.119459  26   
26  1973  145286.0  211536.0  0.9057  11.886459  12.262151 -0.099047  27   
27  1974  112349.0  213743.0  0.9232  11.629365  12.272530 -0.079909  28   
28  1975   95170.0  215973.0  0.9147  11.463420  12.282908 -0.089159  29   
29  1976  117727.0  218280.0  0.9199  11.676124  12.293534 -0.083490  30   
30  1977  142412.0  220612.0  0.9604  11.866480  12.304161 -0.040405  31   
31  1978  151261.0  222968.0  1.0061  11.926762  12.314783  0.006082  32   
32  1979  143049.0  225350.0  1.0356  11.870942  12.325410  0.034981  33   
33  1980  112310.0  227757.0  1.0432  11.629018  12.336035  0.042293  34   
34  1981  102714.0  230139.0  1.0282  11.539703  12.346438  0.027810  35   
35  1982   84676.0  232520.0  1.0000  11.346587  12.356731  0.000000  36   
36  1983  122819.0  234799.0  0.9836  11.718467  12.366485 -0.016536  37   
37  1984  145166.0  237001.0  0.9836  11.885633  12.375819 -0.016536  38   
38  1985  146311.0  239279.0  0.9760  11.893490  12.385386 -0.024293  39   
39  1986  168406.0  241625.0  0.9760  12.034133  12.395143 -0.024293  40   
40  1987  167459.0  243934.0  0.9892  12.028494  12.404653 -0.010859  41   
41  1988  165459.0  246329.0  0.9864  12.016479  12.414423 -0.013693  42   

       invpc    linvpc  lprice_1  linvpc_1    gprice    ginvpc  
0   0.380667 -0.965831       NaN       NaN       NaN       NaN  
1   0.441360 -0.817895 -0.199671 -0.965831  0.054530  0.147935  
2   0.423291 -0.859694 -0.145141 -0.817895 -0.022567 -0.041799  
3   0.567060 -0.567289 -0.167709 -0.859694  0.035890  0.292405  
4   0.457654 -0.781643 -0.131819 -0.567289  0.006142 -0.214353  
5   0.436905 -0.828039 -0.125677 -0.781643  0.002605 -0.046397  
6   0.443819 -0.812338 -0.123072 -0.828039  0.002936  0.015701  
7   0.483155 -0.727418 -0.120136 -0.812338 -0.031036  0.084920  
8   0.551832 -0.594512 -0.151172 -0.727418  0.012829  0.132906  
9   0.477842 -0.738476 -0.138343 -0.594512  0.013800 -0.143964  
10  0.432290 -0.8386

In [2]:

# Calculate first order autocorrelation
print("ACF for linvpc:")
linvpc_acf = acf(hseinv['linvpc'], nlags=1)
print(f"1: {linvpc_acf[1]:.3f}")

print("ACF for lprice:")
lprice_acf = acf(hseinv['lprice'], nlags=1)
print(f"1: {lprice_acf[1]:.3f}")



ACF for linvpc:
1: 0.594
ACF for lprice:
1: 0.896


In [3]:
# Linearly detrend log(invpc)
hseinv['time'] = np.arange(1, len(hseinv) + 1)
lm_invpc = sm.OLS(hseinv['linvpc'], sm.add_constant(hseinv['time'])).fit()
hseinv['detrended_log_invpc'] = lm_invpc.resid

# Linearly detrend log(price)
lm_price = sm.OLS(hseinv['lprice'], sm.add_constant(hseinv['time'])).fit()
hseinv['detrended_log_price'] = lm_price.resid

# Compute first order autocorrelation after detrending
print("ACF for detrended_log_invpc:")
detrended_log_invpc_acf = acf(hseinv['detrended_log_invpc'], nlags=1)
print(f"1: {detrended_log_invpc_acf[1]:.3f}")

print("ACF for detrended_log_price:")
detrended_log_price_acf = acf(hseinv['detrended_log_price'], nlags=1)
print(f"1: {detrended_log_price_acf[1]:.3f}")



ACF for detrended_log_invpc:
1: 0.476
ACF for detrended_log_price:
1: 0.818


### Conclusion for C1 (i)
- even after detrending, the estimated coefficients are still pretty large. Hence, we cannot confidently rule out a unit root in log(price)


In [4]:
# (ii) Estimate the equation
print("\n#(ii) Estimate the equation")
hseinv['diff_log_price'] = hseinv['lprice'].diff()
model1 = sm.OLS(hseinv['linvpc'].iloc[1:], 
               sm.add_constant(hseinv[['diff_log_price', 'time']].iloc[1:])).fit()
print(model1.summary())




#(ii) Estimate the equation
                            OLS Regression Results                            
Dep. Variable:                 linvpc   R-squared:                       0.510
Model:                            OLS   Adj. R-squared:                  0.484
Method:                 Least Squares   F-statistic:                     19.77
Date:                Fri, 18 Apr 2025   Prob (F-statistic):           1.31e-06
Time:                        16:36:39   Log-Likelihood:                 30.090
No. Observations:                  41   AIC:                            -54.18
Df Residuals:                      38   BIC:                            -49.04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const          

### C1(ii)
- the coef on Delta log(price) implies that a 1% increase in price leads to a 3.88% increase in invpc


In [5]:
# (iii) Use detrended log(invpc) as dependent variable

print("\n#(iii) Use detrended log(invpc) as dependent variable")
model2 = sm.OLS(hseinv['detrended_log_invpc'].iloc[1:], 
               sm.add_constant(hseinv[['diff_log_price', 'time']].iloc[1:])).fit()
print(model2.summary())




#(iii) Use detrended log(invpc) as dependent variable
                             OLS Regression Results                            
Dep. Variable:     detrended_log_invpc   R-squared:                       0.303
Model:                             OLS   Adj. R-squared:                  0.266
Method:                  Least Squares   F-statistic:                     8.242
Date:                 Fri, 18 Apr 2025   Prob (F-statistic):            0.00106
Time:                         16:36:40   Log-Likelihood:                 30.090
No. Observations:                   41   AIC:                            -54.18
Df Residuals:                       38   BIC:                            -49.04
Df Model:                            2                                         
Covariance Type:             nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------

### C1(iii)
- after we detrend the log(invpc) before regressing it on delta log(price) and time trend, the R^2 bcomes 0.303, which is lower than the situation before detrending. 
- Hence, delta log(price) explains only 30.3% of the variation in log(invpc) about its trend 

In [6]:
# (iv) Use first difference of log(invpc) as dependent variable

print("\n#(iv) Use first difference of log(invpc) as dependent variable")
hseinv['diff_log_invpc'] = hseinv['linvpc'].diff()
model3 = sm.OLS(hseinv['diff_log_invpc'].iloc[1:], 
               sm.add_constant(hseinv['diff_log_price'].iloc[1:])).fit()
print(model3.summary())



#(iv) Use first difference of log(invpc) as dependent variable
                            OLS Regression Results                            
Dep. Variable:         diff_log_invpc   R-squared:                       0.047
Model:                            OLS   Adj. R-squared:                  0.023
Method:                 Least Squares   F-statistic:                     1.944
Date:                Fri, 18 Apr 2025   Prob (F-statistic):              0.171
Time:                        16:36:41   Log-Likelihood:                 22.986
No. Observations:                  41   AIC:                            -41.97
Df Residuals:                      39   BIC:                            -38.55
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------

### C1(iv)
- the coef on delta log(price) is no longer significant at 5 % level. 
- The R^2 is much smaller. 
- delta log(price) explains very little variation in delta log (invpc). 
- Since differencing eliminates linear time trends, hence the result is not surprising to be insignificant.

In [7]:
# (C4)
print("\n#(C4)")
print("#(i)Estimate a Phillips curve where inflation and unemployment are in first differences")

# Load phillips dataset
phillips = wooldridge.data('phillips')

# Create time series data starting from 1948
phillips.index = pd.date_range(start='1948', periods=len(phillips), freq='A')

# Calculate first differences
phillips['d_inf'] = phillips['inf'].diff()
phillips['d_unem'] = phillips['unem'].diff()

# Filter data up to 2003
phillips_2003 = phillips[phillips.index.year <= 2003]

# Estimate the model
reg_ea = sm.OLS(phillips_2003['d_inf'].dropna(), 
               sm.add_constant(phillips_2003['d_unem'].dropna())).fit()
print(reg_ea.summary())




#(C4)
#(i)Estimate a Phillips curve where inflation and unemployment are in first differences
                            OLS Regression Results                            
Dep. Variable:                  d_inf   R-squared:                       0.135
Model:                            OLS   Adj. R-squared:                  0.118
Method:                 Least Squares   F-statistic:                     8.256
Date:                Fri, 18 Apr 2025   Prob (F-statistic):            0.00583
Time:                        16:36:42   Log-Likelihood:                -122.03
No. Observations:                  55   AIC:                             248.1
Df Residuals:                      53   BIC:                             252.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------

/var/folders/kz/g7fm_pz11tv5kbttdpy6thc40000gn/T/ipykernel_19787/1205131528.py:9: FutureWarning: 'A' is deprecated and will be removed in a future version, please use 'YE' instead.
  phillips.index = pd.date_range(start='1948', periods=len(phillips), freq='A')


### C4(i)
- the const is not significant, while d_unem is significant even at 0.01 level. 
- Both sign of d_unem and const are negative, which is consistent with the logic of Phillips curve.
- R^2 = 0.135


In [8]:
# (ii) Comparison of both Phillips curve models

print("\n#(ii)Comparison of both Phillips curve models")

reg_ea2 = sm.OLS(phillips_2003['d_inf'].dropna(), 
                sm.add_constant(phillips_2003['unem'].iloc[1:])).fit()
print(reg_ea2.summary())


#(ii)Comparison of both Phillips curve models
                            OLS Regression Results                            
Dep. Variable:                  d_inf   R-squared:                       0.104
Model:                            OLS   Adj. R-squared:                  0.087
Method:                 Least Squares   F-statistic:                     6.132
Date:                Fri, 18 Apr 2025   Prob (F-statistic):             0.0165
Time:                        16:36:43   Log-Likelihood:                -123.00
No. Observations:                  55   AIC:                             250.0
Df Residuals:                      53   BIC:                             254.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const

### C4(ii)
- the parameters are less significant compare to the first model, however still significant at 0.05 level. And the coef becomes less negative. 
- the R^2 become lesser than the first model. 
- Hence, the first model is preferred, rather than the second model. 



In [9]:
# (C8)
print("\n#(C8)")
print("#(i)Obtain AR(1) for unemployment rate and make a prediction for that of 2004.")

# Create lag variables
phillips['unem_1'] = phillips['unem'].shift(1)
phillips['inf_1'] = phillips['inf'].shift(1)

# Filter data up to 2003
phillips_2003 = phillips[phillips.index.year <= 2003]

# AR(1) model for unemployment
model1 = sm.OLS(phillips_2003['unem'].iloc[1:], 
               sm.add_constant(phillips_2003['unem_1'].iloc[1:])).fit()

# Make prediction for 2004
# Get the last unemployment rate (2003) to predict 2004
unem_2003 = phillips[phillips.index.year == 2003]['unem'].values[0]
pred_data = pd.DataFrame({'const': [1], 'unem_1': [unem_2003]})
f1 = model1.predict(pred_data)
print(f"Prediction for 2004 using AR(1): {f1.values[0]:.6f}")

actual_2004 = 5.5  # The actual 2004 unemployment rate
print(f"Actual 2004 unemployment rate: {actual_2004:.1f}%")
print(f"Prediction error: {f1.values[0] - actual_2004:.6f}")




#(C8)
#(i)Obtain AR(1) for unemployment rate and make a prediction for that of 2004.
Prediction for 2004 using AR(1): 5.943979
Actual 2004 unemployment rate: 5.5%
Prediction error: 0.443979


In [10]:
# (ii) Add lag inflation rate into the model
print("\n#(ii)Add lag inflation rate into the model")
model2 = sm.OLS(phillips_2003['unem'].iloc[1:], 
               sm.add_constant(phillips_2003[['unem_1', 'inf_1']].iloc[1:])).fit()

# Print summary of both models
print("\nModel 1 (AR(1)):")
print(f"Observations: {model1.nobs}")
print(f"Adjusted R-squared: {model1.rsquared_adj:.3f}")
print(f"Residual Std. Error: {np.sqrt(model1.scale):.3f}")
print("\nModel 2 (AR(1) with inflation):")
print(f"Observations: {model2.nobs}")
print(f"Adjusted R-squared: {model2.rsquared_adj:.3f}")
print(f"Residual Std. Error: {np.sqrt(model2.scale):.3f}")




#(ii)Add lag inflation rate into the model

Model 1 (AR(1)):
Observations: 55.0
Adjusted R-squared: 0.558
Residual Std. Error: 0.999

Model 2 (AR(1) with inflation):
Observations: 55.0
Adjusted R-squared: 0.685
Residual Std. Error: 0.843


In [11]:
# (iii) Again make a prediction using modified model
print("\n#(iii)Again make a prediction using modified model")

# Get the last unemployment and inflation rates (2003) to predict 2004
unem_2003 = phillips[phillips.index.year == 2003]['unem'].values[0]
inf_2003 = phillips[phillips.index.year == 2003]['inf'].values[0]
pred_data2 = pd.DataFrame({'const': [1], 'unem_1': [unem_2003], 'inf_1': [inf_2003]})
f2 = model2.predict(pred_data2)
print(f"Prediction for 2004 using modified model: {f2.values[0]:.6f}")
print(f"Actual 2004 unemployment rate: {actual_2004:.1f}%")
print(f"Prediction error: {f2.values[0] - actual_2004:.6f}")




#(iii)Again make a prediction using modified model
Prediction for 2004 using modified model: 5.614208
Actual 2004 unemployment rate: 5.5%
Prediction error: 0.114208


In [20]:
# (iv) Construct a five-year 95% prediction interval for new observations and check the range
print("\n#(iv)Construct a five-year 95% prediction interval for new observations and check the range")

for year in range(2000, 2005):
    # Get the previous year's unemployment rate
    prev_year = year - 1
    unem_prev = phillips[phillips.index.year == prev_year]['unem'].values[0]
    
    # Create prediction data
    pred_data = pd.DataFrame({'const': [1], 'unem_1': [unem_prev]})
    
    # Get prediction with intervals
    prediction = model1.get_prediction(pred_data)
    frame = prediction.summary_frame(alpha=0.05)
    print(f"{year}: fit={frame['mean'].values[0]:.6f}, lwr={frame['obs_ci_lower'].values[0]:.6f}, upr={frame['obs_ci_upper'].values[0]:.6f}")
    



#(iv)Construct a five-year 95% prediction interval for new observations and check the range
2000: fit=4.607690, lwr=2.570527, upr=6.644854
2001: fit=4.459214, lwr=2.417247, upr=6.501181
2002: fit=5.053120, lwr=3.026635, upr=7.079606
2003: fit=5.795503, lwr=3.774204, upr=7.816801
2004: fit=5.943979, lwr=3.921815, upr=7.966143
